Feature/55 create spark metadata

filiphornak opened the pull request:
Implements #55 

 Create _spark_metadata.

* Implemented subcommand that can generate _spark_metadata directory and metadata files up to maxMicroBatchNumber
* Metadata are generated from the last .compact file
* Unintentionally reformatted few files on save
* Still WIP, some tests are not implemented yet, also the tool should produce logs while creating metadata for simpler investigation

Commit:
feat: Create _spark_metadata (#55)

Commit:
refactor: Simplified version of createMetadata function (#55)

Commit:
test: Create placeholders for test functions (#55)

Commit:
refactor: Remove unused imports (#55)

Commit:
tests: Add tests for MetadataTool + HdfsFileManager (#55)

Commit:
tests: Add test for hdfs failure while making dir

Commit:
refactor: Use hdfs FileStatus instead of ToolFileStatus

Commit:
feat: WIP walk partition tree for datafiles.

Commit:
feat: walk partition tree for datafiles.

Commit:
fix: Unix filesystem can generate _spark_metadata

Commit:
test: Add missing tests for s3

Commit:
feat: Set block size to hadoop default and replication to 1

Commit:
refactor: Add appropriate types to Defaults

Commit:
fix (s3): Tool can now properly create _spark_metadata on s3

Commit:
perf (s3): Get FileStatus-es in single request from s3

Commit:
docs (README): Mention `create-metadata` subcommand

filiphornak commented:
It's tested and working, however I might misunderstood the concept of `--max-micro-batch-number`. Mainly how should the tool behave when such number doesn't match number of files in particular directory?

kevinwallimann commented:
> It's tested and working, however I might misunderstood the concept of `--max-micro-batch-number`. Mainly how should the tool behave when such number doesn't match number of files in particular directory?

The number of files doesn't have to be equal to `--max-micro-batch-number`, see example 1 in #55, where there are 10 files, and the max micro batch number is 9. The `--max-micro-batch-number` should be chosen by the user, and the user should choose it to be equal to the latest committed offset in the checkpoints directory. Of course, the user may choose to manually edit the checkpoints directory, so we didn't want to create a hard dependency on the checkpoints directory in this tool.

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/ArgumentParser.scala:
```diff
@@ -92,6 +93,24 @@ object ArgumentParser {
       note(sys.props("line.separator")),
+      cmd("create-metadata")
+        .text("Create Spark structured streaming metadata")
+        .action((_, c) => c.copy(mode = CreateMetadata(maxMicroBatchNumber = 0, compactionNumber =  1)))
```
@jozefbakus I think the default for compaction number should be 10, as it is in Spark. What do you think?

filiphornak replied:
I changed it, as you suggested. However It's also an required option. This might be changed to just optional, what do you think?

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/io/UnixFileManager.scala:
```diff
@@ -68,10 +69,33 @@ case object UnixFileManager extends FileManager {
+    paths.foreach(p => Files.delete(Paths.get(toUriWithScheme(p))))
   }.toEither.map(_ => ()).leftMap(err => IoError(err.getMessage, err.some))
 
+  override def walkFileStatuses(baseDir: Path, filter: Path => Boolean): Either[IoError, Seq[FileStatus]] =
```
You forgot to use the `filter`, and the sorting

filiphornak replied:
Thank you! I didn't notice that.

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/DataTool.scala:
```diff
@@ -0,0 +1,50 @@
+  import za.co.absa.spark_metadata_tool.DataTool._
+
+  def listStatusesUpToPart(path: Path, maxPartNumber: Int): Either[IoError, Seq[FileStatus]] =
+    listDataFileStatuses(path).map(_.take(maxPartNumber))
```
As discussed: Take all file statuses

filiphornak replied:
Yeah, I changed the implementation. So in current implementation it's guaranteed that at leas last loose metadata file contains any metadata entries, or in other words the sequence of metadata is aligned to highest metadata file. As we discussed before, spark actually doesn't care about the content of the metadata files much, only about that it starts with `v1` string and json records, if any, are valid.

If there is requirement, to store them in other specific way, we could easily extend it with something like strategy:
```scala
case class Metadata(batchNum: Int, statuses: Seq[SinkFileStatus])

trait MetadataStoreStrategy {
  def compactedMetadata(metadata: Seq[SinkFIleStatus]): Option[Metadata]

  def looseMetadata(metadata: Seq[SinkFileStatus]): Seq[Option[Metadata]]
}

object MedataStoreStrategy {
  final class MaxBatchAligned(...) extends MetadataStoreStrategy {
    // impl
  }

  final class InCompactOrFirst(...) extends MetadataStoreStrategy {
    // impl
  }

  // other types
}
```

Or just change the implementation 😄 .

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/io/S3FileManager.scala:
```diff
@@ -101,39 +101,61 @@ case class S3FileManager(s3: S3Client) extends FileManager {
+        obj    <- page.contents().asScala
+        status <- Some(new Path(builder.setPath(obj.key()).build())).filter(filter).map(asFileStatus(obj))
+      } yield status
+    }.map(_.toSeq)
```
Maybe add a comment here, that ListObjectsV2 returns the objects sorted by key ascending.

filiphornak replied:
👍

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/io/FileManager.scala:
```diff
@@ -26,4 +26,6 @@ trait FileManager {
   def write(path: Path, lines: Seq[String]): Either[IoError, Unit]
   def copy(from: Path, to: Path): Either[IoError, Unit]
   def delete(paths: Seq[Path]): Either[IoError, Unit]
+  def makeDir(dir: Path): Either[IoError, Unit]
```
Please add a comment that makeDir will create a directory only if it doesn't exist, and only if its parent directory exists.

filiphornak replied:
I added docstring for those methods in trait. The docstring should be generic enugh, that it points out only the common specified behavior, like the validation of existence of parent directory and check that file with same name doesn't exist.

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/io/FileManager.scala:
```diff
@@ -26,4 +26,6 @@ trait FileManager {
   def copy(from: Path, to: Path): Either[IoError, Unit]
   def delete(paths: Seq[Path]): Either[IoError, Unit]
+  def makeDir(dir: Path): Either[IoError, Unit]
+  def walkFileStatuses(baseDir: Path, filter: Path => Boolean): Either[IoError, Seq[FileStatus]]
```
Please make a comment here about the sort order

kevinwallimann commented on src/test/scala/za/co/absa/spark_metadata_tool/io/HdfsFileManagerSpec.scala:
```diff
@@ -212,5 +223,129 @@ class HdfsFileManagerSpec extends AnyFlatSpec with Matchers with OptionValues wi
+  it should "fail when hdfs FileSystem failure occurs" in {
+    val parentPath = new Path("hdfs:///root/")
+    val dirPath    = new Path(parentPath, "new")
+    val fsMock = new FileSystem {
```
Out of curiosity, is there a specific reason why you chose to provide a fake implementation of `FileSystem` here instead of a `mock[FileSystem]`?

filiphornak replied:
Because I couldn't actually create mock. It's because this is not `trait` (or `interface`, or `purely virtual class` and it contains some inner classes and methods that couldn't be mocked. We actually talked about it with @jozefbakus, because I wanted to create `IFileSystem` proxy trait, however as Jozef pointed out, it would be unnecessary overkill for one test.

kevinwallimann commented on src/test/scala/za/co/absa/spark_metadata_tool/CompactionSpec.scala:
```diff
@@ -0,0 +1,41 @@
+  it should "return None for small micro batch number (no compacted files were created)" in {
+    val maxUBatchNum = 0
+    val compactionNum = 12
+    Compaction.lastCompaction(maxUBatchNum, compactionNum) should equal(None)  // Maybe Option[Int] would be nice
```
You have `Option[Int]`, don't you? Did you forget to remove this comment?

filiphornak replied:
Removed that comment. It was forgotten left over, before change of method signature. Before `lastCompaction` returned just `Int`. In situations without compaction file it returned `-1`, what is result of `maxMicroBatchNum - (maxMicroBatchNum % compactionPeriod) - 1`.

kevinwallimann commented on src/test/scala/za/co/absa/spark_metadata_tool/CompactionSpec.scala:
```diff
@@ -0,0 +1,41 @@
+
+class CompactionSpec extends AnyFlatSpec with Matchers {
+  "lastCompaction" should "calculate number of last compaction file" in {
+    val maxUBatchNum = 20
```
Does U stand for Micro? Laziness is the engine for progress 😄 , but please write it out `maxMicrobatchNum`

filiphornak replied:
Yeah, U stands for micro in this context. I changed it as you asked, however what about `µ` 😄?

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/ArgumentParser.scala:
```diff
@@ -92,6 +93,24 @@ object ArgumentParser {
+          opt[Int]('m', "max-micro-batch-number")
+            .required()
+            .action((x, c) => c.copy(mode = c.mode.asInstanceOf[CreateMetadata].copy(maxMicroBatchNumber = x)))
+            .text("set max batch number"),
```
If you pass `--compaction-number 0`, then you get a `java.lang.ArithmeticException: / by zero` in `Compaction.scala`. It would be nicer to validate the input

filiphornak replied:
Good find, thank you! I added validation, that the number must be positive. I think maybe I should also add this validation to `--max-micro-batch-number`.

kevinwallimann commented on src/main/scala/za/co/absa/spark_metadata_tool/model/Action.scala:
```diff
@@ -0,0 +1,22 @@
+
+object Action {
+  val Add: String    = "add"
+  val Delete: String = "delete"
```
`Delete` is unused.

filiphornak replied:
I removed it. Thank you.

kevinwallimann reviewed (commented):
Very nice code!

Commit:
fix (unix): use filter and sortBy file uri.

UnixFileManager#walkFileStatuses now utilizes filter function and sorts output by path uri

Commit:
fix (unix): Set default for compactionNumber to 10

Commit:
fix (s3): Add comment about ordering in listObjects

Commit:
docs (FileManager): Add docstring to newly created methods.

Description was added to interface to better describe their intended behavior and behavior of their implementations.

Commit:
refactor (Action): Remove Delete action.

Commit:
fix (ArgumentParser): Add validation that compaction number is positive

Commit:
fix (createMetadata): Changed behavior of splitting status files between metadata files

jozefbakus commented on src/main/scala/za/co/absa/spark_metadata_tool/io/FileManager.scala:
```diff
@@ -26,4 +26,30 @@ trait FileManager {
+    *
+    * Note that directory will be created only if following conditions are met:
+    *
+    *   - Directory will be created only if parent directory
```
Unfinished sentence: "Directory will be created only if parent directory exists"

filiphornak replied:
Thank you! I overlooked that.

jozefbakus commented on src/main/scala/za/co/absa/spark_metadata_tool/io/HdfsFileManager.scala:
```diff
@@ -65,7 +66,29 @@ case class HdfsFileManager(hdfs: FileSystem) extends FileManager {
-  private def listDirectory(path: Path): Either[IoError, Seq[Path]] = {
+  override def makeDir(dir: Path): Either[IoError, Unit] =
+    for {
+      fileExists   <- checkDirectoryExists(dir)
```
checkDirectoryExists but variable name is file. The same in IoError bellow.

Also I think method checkDirectoryExists is not enough for this one, because it only returns if dir exists but there might be a file with the same name

filiphornak replied:
You are right, I'll fix that one.

filiphornak replied:
I thought about it for a few moments and came up to a conclusion about why I didn't implement it that way. It's because if there is a file with the same name as a dir name. Hdfs would throw an exception or ignore it because if there might be files with the same names as directories. This implementation prevents `hdfs.makeDir` from creating more than one child at a time. However, I already fixed it to check if there is no other resource with the same name.

jozefbakus commented on src/main/scala/za/co/absa/spark_metadata_tool/io/UnixFileManager.scala:
```diff
@@ -90,5 +114,8 @@ case object UnixFileManager extends FileManager {
   }.fold(err => Left(IoError(err.getMessage, err.some)), res => Right(res))
 
-  private def withScheme(path: Path): Path = new Path(s"file:///${path.toString}")
+  private def toUriWithScheme(path: Path): URI = {
```
I like this "refactoring"

jozefbakus commented on src/main/scala/za/co/absa/spark_metadata_tool/io/S3FileManager.scala:
```diff
@@ -101,39 +101,62 @@ case class S3FileManager(s3: S3Client) extends FileManager {
+    // listObjectsV2 returns objects in ascending order by key, so we don't have to order them
+    catchAsIoError {
+      for {
+        page   <- s3.listObjectsV2Paginator(request).iterator().asScala
```
Is this recursive ?

filiphornak replied:
Yes, it will list all objects on s3 with this prefix. As you might have noticed, there isn't a delimiter set for the request as in the `listBucket` method, that's why it appears to walk recursively through the keys.

Commit:
fix: Docs + hdfs mkDir logic after review

jozefbakus reviewed (approved):
Nice👍
